### PR TITLE
Feature/protocol tests

### PR DESCRIPTION
**Pull request by joscas**

### CODE REVIEW

File: domain/src/main/scala/org/parsec/protocol/Parsec.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+      if (currency == "ETH")
+        return buyerAddress equals ethereumAddressFromPublicKey(buyerPublicKey)
+      else
+        return ((buyerAddress equals bitcoinP2PKHAddressFromPublicKey(buyerPublicKey))
+               ||
+               (buyerAddress equals bitcoinP2SHAddressFromPublicKey(buyerPublicKey)))
+    }
+  }
+
+  def ethereumAddressFromPublicKey(key: String): String = {
```
Comment (Antwnis) [inline domain/src/main/scala/org/parsec/protocol/Parsec.scala]: I think we should introduce one or more `traits` EthereumEncryption BitcoinEncryption etc - to allow mix-in of capabilities

Comment (joscas) [reply domain/src/main/scala/org/parsec/protocol/Parsec.scala]: good idea

**Comment by joscas** [comment]
Please check now @Antwnis

File: build.sbt [hunk 2]
```diff
@@ -74,7 +79,9 @@
     "com.typesafe" % "config" % "1.2.1",
     slf4jApi,
     log4jOverSlf4j,
-    logbackClassic
+    logbackClassic,
+    bouncycastle,
```
Comment (pauloglesby) [inline build.sbt]: I'd use `baseJavaDependencies` just for java libraries that could also be used in java-only modules. The `crypto` dependencies can be added where we need them, e.g. in the domain module we currently have:
````scala
lazy val domain = Project(
     "domain",
     file("domain"),
     settings = scalaProjectSettings ++ Seq(
       libraryDependencies ++= testing
     )
   ).dependsOn(common % fullScope)
````
Which could be:
````scala
lazy val domain = Project(
     "domain",
     file("domain"),
     settings = scalaProjectSettings ++ Seq(
       libraryDependencies ++= testing ++ crypto
     )
   ).dependsOn(common % fullScope)
````

Comment (joscas) [reply build.sbt]: Good suggestion. Let me update

**Comment by joscas** [comment]
Check now @pauloglesby

File: domain/src/main/scala/BitcoinCryptography.scala [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package org.parsec.protocol
+
+trait BitcoinCryptography {
```
Comment (pauloglesby) [inline domain/src/main/scala/BitcoinCryptography.scala:3]: We could have something like:
````scala
/**
 * somewhere in the domain, and replacing the existing `string` types for keys in the protocol
 */
sealed trait PublicKey
case class BTCPublicKey(value: String) extends PublicKey {
  def p2pkhAddress = ???
  def p2shAddress = ???
  // plus validation methods...
}
case class ETHPublicKey(value: String) extends PublicKey {
  def address = ???
  // plus validation methods
}
````

That would let us keep all the methods with the data, and start moving us away from `String`s to value types (which will open up a lot more FP power as the project grows).

File: build.sbt [hunk 4]
```diff
@@ -165,7 +174,7 @@
     "domain",
     file("domain"),
     settings = scalaProjectSettings ++ Seq(
-      libraryDependencies ++= testing
+      libraryDependencies ++= testing ++ crypto
```
Comment (pauloglesby) [inline build.sbt:177]: 👍

File: domain/src/main/scala/EthereumCryptography.scala [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import org.apache.commons.codec.binary.Hex
+import org.bouncycastle.jcajce.provider.digest.Keccak
+
+trait EthereumCryptography {
+  def ethereumAddressFromPublicKey(key: String): String = {
+    val keyBytes = Hex.decodeHex(key.toCharArray) // Convert HEX public key to bytes
+    val keccak256 = new Keccak.Digest256
+    val keyHash = keccak256.digest(keyBytes) // Take the Keccak-256 hash of the public key
+    val addressBytes = util.Arrays.copyOfRange(keyHash, 12, keyHash.length) // Drop the 12 first bytes of the hash
+    Hex.encodeHexString(addressBytes) // Return the hex hash
```
Comment (pauloglesby) [inline domain/src/main/scala/EthereumCryptography.scala:13]: Nice!

File: domain/src/main/scala/BitcoinCryptography.scala (hunk 3 above)
Comment (joscas) [reply domain/src/main/scala/BitcoinCryptography.scala:3]: Agreed but let us implement this after a couple of PRs I have in store which will clarify the domain further?

Comment (pauloglesby) [reply domain/src/main/scala/BitcoinCryptography.scala:3]: Perfect - I'm happy to merge now. We can put a `TODO` in for that.

**Pull request merged**